src/javascript/gameInstance.js

```diff
@@ -116,69 +130,214 @@ export default class GameInstance {
+              this.onKeyDown(track.keyBind);
+              touches[pointer.identifier] = track.keyBind;
+            } else {
+              this.onKeyUp(track.keyBind);
+              touches[pointer.identifier] = null;
+            }
           }
         });
       }
     };
 
-    this.touchRegion = ZingTouch.Region(this.canvas);
+    const moveEvent = (e) => {
+      for (let pointer of e.changedTouches) {
+        const x = pointer.clientX;
+        if (!touches[pointer.identifier]) return;
 
-    for (let numInputs of [1, 2, 3, 4]) {
-      this.touchRegion.bind(
-        this.canvas,
-        new ZingTouch.Tap({ numInputs }),
-        tapEvent
-      );
-    }
+        this.dropTrackArr.forEach((track) => {
+          if (x > track.x && x < track.x + track.width) {
+            if (touches[pointer.identifier] !== track.keyBind) {
+              this.onKeyUp(touches[pointer.identifier]);
+              touches[pointer.identifier] = null;
+            }
+          }
+        });
+      }
+    };
+
+    this.canvas.ontouchstart = (e) => {
+      tapEvent(e, true);
+    };
+    this.canvas.ontouchmove = moveEvent;
+    this.canvas.ontouchcancel = (e) => {
+      tapEvent(e, false);
+    };
+    this.canvas.ontouchend = (e) => {
+      tapEvent(e, false);
+    };
   }
 
   // log key and touch events
-  async onKeyDown(key) {
+  async onKeyDown(keyRaw) {
+    const key = keyRaw.toLowerCase();
+    // avoid repeated triggering when key is held
+    if (this.keyStatus[key]) return;
+    this.keyStatus[key] = true;
+    if (!this.trackKeyBind.includes(key)) return;
+    // if in create mode, create note
     if (!this.vm.playMode && !this.paused) {
       const cTime = await this.getCurrentTime();
-      if (this.trackKeyBind.includes(key)) {
-        // this.timeArr.push({ t: cTime.toFixed(3), k: key });
-        // this.timeArrIdx = this.timeArr.length - 1;
-        if (this.lastAddedTime && cTime - this.lastAddedTime < 0.05) {
-          this.timeArr[this.lastAddedIdx].k += key;
-        } else {
-          // add at idx
-          this.timeArr.splice(this.timeArrIdx, 0, {
-            t: Number(cTime.toFixed(3)),
-            k: key,
-          });
-          this.lastAddedTime = cTime;
-          this.lastAddedIdx = this.timeArrIdx;
+      this.createSingleNote(key, cTime);
+      const singleNoteObj = this.timeArr[this.lastAddedIdx];
+      // convert to hold note
+      setTimeout(() => {
+        if (
+          this.keyStatus[key] &&
+          singleNoteObj == this.timeArr[this.lastAddedIdx]
+        ) {
+          this.createHoldNote(key, singleNoteObj);
         }
-      }
+      }, 300);
+    }
+    // register key down
+    for (const track of this.dropTrackArr) {
+      track.keyDown(key);
+    }
+  }
+
+  async onKeyUp(key) {
+    this.keyStatus[key] = false;
+    if (this.holdingNote[key]) {
+      const cTime = await this.getCurrentTime();
+      this.holdingNote[key].h[key] = cTime;
+      this.holdingNote[key] = null;
```

Comment:
Is this intended?

Reply:
Yes, I've added some comment to clarify.